Automattic/wp-calypso
# Checkout: add unit tests for payment processors

#### Changes proposed in this Pull Request

This PR adds basic functionality unit tests to all the payment processor functions.

In order for this to work, we needed to make sure that all the data used by the processors comes from their arguments, so we had to add the contact details data to the options passed to every processor. Previously every processor pulled the contact details from the `wpcom` data store itself, but this is very difficult to mock and I wasn't able to find a way to make it work. This change could be moved into its own PR once all these tests are complete and the tests themselves can prove that it works.

Note: This does not currently create tests for the 3DS authentication flows of `multiCardProcessor` and `existingCardProcessor` because I can't easily mock `confirmStripePaymentIntent` and that function directly accesses the `window.Stripe` object. It's possible to create tests there but I think it's of lower importance and they can be added later.

- [x] multiCardProcessor
- [x] existingCardProcessor
- [x] genericRedirectProcessor
- [x] freePurchaseProcessor
- [x] fullCreditsProcessor
- [x] payPalExpressProcessor
- [x] applePayProcessor
- [x] weChatProcessor

#### Testing instructions

Make sure tests pass.

```
yarn run test-client client/my-sites/checkout/composite-checkout/test/
```
